### PR TITLE
Branches-Steph

### PR DESCRIPTION
**Pull request by Steph0088**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is where all code that active records uses to interact with the database. 
Describe in your own words what the Controller is doing in Rails | THe controller sends requests to the correct place in order to return the users request. 
Describe in your own words what the View is doing in Rails | The view is taking the data to complete the users request and acting on it to represent it in a way the user will find useful. 
Describe an edge-case controller test you wrote | I don't think I wrote any edge-case tests. 
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params are used as  a measure of security and a way to simply the creating of instances or objects. 
How are Rails migrations related to Rails models? | Migrations are used to make changes to the model data schema. 
Describe one area of Rails that are still unclear on | I am still confused on the syntax, what goes where, mostly symbols and key value pairs, when to use them as well as params. What is allowed and what isn't.

### CODE REVIEW

**Comment by tildeee** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | x
Answered comprehension questions | x
Successfully handles: Index, Show | x
Index & Show tests pass | x
Successfully handles: New, Create | x
New & Create tests pass | x
Successfully handles: Edit, Update | x
Tests for Edit & Update test valid & invalid task ids | x
Successfully handles: Destroy, Task Complete | x
Tests for Destroy & Task Complete include tests for valid and invalid task ids | x
Routes follow RESTful conventions | x
Uses named routes (like `_path`) | x
**Overall** |



Fantastic work on this project, Steph! Your project looks so so so good! The code is all there and works well, and your tests are all there and work well! I have no complaints about this project


I only have a few comments that are suggestions on how to refactor the code, and also some few ways to make your code more Rails-y.


Lastly, I may say that your TaskList webpages didn't have a direct link to the Task's show page, and in the future, that would be convenient for me :P


Otherwise, great work on this, and keep it up!

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#TASKS = ["Laundry", "Feed Babette", "Clean room", "Sleep"] 
+class TasksController < ApplicationController
+  
+  def index
+    @tasks = Task.all
+  end
+  
+  def show
+    task_id = params[:id].to_i
+    @task = Task.find_by(id: params[:id])
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:10]: Consider refactoring the above two lines to either

```ruby
    task_id = params[:id].to_i
    @task = Task.find_by(id: task_id)
```

or

```ruby
    @task = Task.find_by(id: params[:id])
```

in order to help efficiently use local variables

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      return
+    end
+  end
+  
+  def new
+    @task = Task.new
+  end
+  
+  def create
+    @task = Task.new(name: params[:task][:name], description: params[:task][:description], completed: params[:task][:completed])
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:22]: In the future, let's use strong params, so refactor this line to:

```ruby
@task = Task.new(task_params)
```

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    if @task.save
+      redirect_to task_path(@task.id)
+    else
+      redirect_to root_path
+    end
+  end 
+  
+  def edit
+    task_id = params[:id].to_i
+    @task = Task.find_by(id: params[:id] )
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:32]: Same as my comments above in the `show` action, consider refactoring these two lines

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      redirect_to tasks_path
+      return
+    end
+  end
+  
+  def update
+    @task = Task.find_by(id: params[:id])
+    
+    if !@task.nil?
+      @task.update(task_params)
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:44]: Nice use of strong params here!

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    else
+      the_correct_task.destroy
+      redirect_to root_path
+      return
+    end
+  end
+
+  def toggle_complete
+    task_id = params[:id].to_i
+    @task = Task.find_by(id: params[:id] )
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:66]: Same as above-- consider refactoring to use `task_id` or to get rid of it :)

File: db/seeds.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# This file should contain all the record creation needed to seed the database with its default values.
+# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
+#
+# Examples:
+#
+#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
+#   Character.create(name: 'Luke', movie: movies.first)
+raic
```
Comment (tildeee) [inline db/seeds.rb:8]: Nitpick: This stray code was left here and was broken. Watch out!

File: test/controllers/tasks_controller_test.rb [hunk 7]
```diff
@@ -3,124 +3,191 @@
...
+          completed: Date.today
+        }
+      }
+      #Act:
+      #actually update the book data with the new hash form data you created.
+      patch task_path(existing_task.id), params: updated_task_form_data
+      
+      # Assert
+      #Step Three once you change the fields, check to see if they are different than original 
+      #expect(Book.find_by(id: existing_book.id).title).must_equal "Name that you changed book to"
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:136]: Not sure if this would help, but maybe if you can spend the time to remove these comments before submission, or find a way to reword them, that may help with solidfying your understanding

File: test/controllers/tasks_controller_test.rb [hunk 8]
```diff
@@ -3,124 +3,191 @@
...
-    # Your tests go here
+    
+    it "successfully marks a task as complete with completed date" do
+      #Arrange
+      #when the user clicks on the complete button, the program will go back into the data base and
+      #change the completed date from nil and set it to the current date.
+      
+      existing_task =  Task.create(name: "cs fun", description: "recursion practice", completed: nil)
+      patch complete_path(existing_task)
+      expect(Task.find_by(id: existing_task.id).completed).must_be_instance_of  ActiveSupport::TimeWithZone
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:182]: Nice! It may look weird, but it makes sense to me and it's what works.

Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:182]: This test would be perfect and complete with an assertion to check the response, like `must_redirect_to` or something
